pkg/api/validate/dynamic/dynamic.go

```diff
@@ -292,34 +297,95 @@ func (dv *dynamic) validateNatGatewayPermissions(ctx context.Context, s Subnet)
-	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	c := closure{dv: dv, ctx: ctx, resource: r, actions: actions}
+	conditionalFunc := c.usingListPermissions
+	timeout := 5 * time.Minute
+	if dv.pdpClient != nil {
+		conditionalFunc = c.usingCheckAccessV2
+		timeout = 90 * time.Second
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, timeout)
 	defer cancel()
 
-	return wait.PollImmediateUntil(20*time.Second, func() (bool, error) {
-		dv.log.Debug("retry validateActions")
-		perms, err := dv.permissions.ListForResource(ctx, r.ResourceGroup, r.Provider, "", r.ResourceType, r.ResourceName)
+	return wait.PollImmediateUntil(timeout, conditionalFunc, timeoutCtx.Done())
+}
 
-		if detailedErr, ok := err.(autorest.DetailedError); ok &&
-			detailedErr.StatusCode == http.StatusForbidden {
-			return false, steps.ErrWantRefresh
-		}
-		if err != nil {
+// closure is the closure used in PollImmediateUntil's ConditionalFunc
+type closure struct {
+	dv       *dynamic
+	ctx      context.Context
+	resource *azure.Resource
+	actions  []string
+}
+
+// usingListPermissions is how the current check is done
+func (c closure) usingListPermissions() (bool, error) {
+	c.dv.log.Debug("retry validateActions with ListPermissions")
+	perms, err := c.dv.permissions.ListForResource(c.ctx, c.resource.ResourceGroup, c.resource.Provider, "", c.resource.ResourceType, c.resource.ResourceName)
+
+	if detailedErr, ok := err.(autorest.DetailedError); ok &&
+		detailedErr.StatusCode == http.StatusForbidden {
+		return false, steps.ErrWantRefresh
+	}
+	if err != nil {
+		return false, err
+	}
+
+	for _, action := range c.actions {
+		ok, err := permissions.CanDoAction(perms, action)
+		if !ok || err != nil {
+			// TODO(jminter): I don't understand if there are genuinely
+			// cases where CanDoAction can return false then true shortly
+			// after. I'm a little skeptical; if it can't happen we can
+			// simplify this code.  We should add a metric on this.
 			return false, err
 		}
+	}
+	return true, nil
+}
 
-		for _, action := range actions {
-			ok, err := permissions.CanDoAction(perms, action)
-			if !ok || err != nil {
-				// TODO(jminter): I don't understand if there are genuinely
-				// cases where CanDoAction can return false then true shortly
-				// after. I'm a little skeptical; if it can't happen we can
-				// simplify this code.  We should add a metric on this.
-				return false, err
-			}
+// usingCheckAccessV2 uses the new RBAC checkAccessV2 API
+func (c closure) usingCheckAccessV2() (bool, error) {
+	c.dv.log.Debug("retry validationActions with CheckAccessV2")
+	oid, err := aad.GetObjectId(c.dv.authorizer.OAuthToken())
+	if err != nil {
+		return false, err
+	}
+
+	authReq := createAuthorizationRequest(oid, c.resource.String(), c.actions...)
+	results, err := c.dv.pdpClient.CheckAccess(c.ctx, authReq)
+	if err != nil {
+		return false, err
+	}
+
+	for _, result := range results.Value {
```

Comment:
The return value for `CheckAccess` is a pointer to a struct, so this may lead to nil pointer dereference issues. Consider checking if `results` is nil before accessing it.